src/coreclr/jit/hwintrinsicarm64.cpp

```diff
@@ -1746,22 +1747,71 @@ GenTree* Compiler::impSpecialIntrinsic(NamedIntrinsic        intrinsic,
+        case NI_AdvSimd_StoreSelectedScalar:
+        case NI_AdvSimd_Arm64_StoreSelectedScalar:
         {
-            assert(sig->numArgs == 2);
             assert(retType == TYP_VOID);
 
-            CORINFO_ARG_LIST_HANDLE arg1     = sig->args;
-            CORINFO_ARG_LIST_HANDLE arg2     = info.compCompHnd->getArgNext(arg1);
-            var_types               argType  = TYP_UNKNOWN;
-            CORINFO_CLASS_HANDLE    argClass = NO_CLASS_HANDLE;
+            CORINFO_ARG_LIST_HANDLE arg1                = sig->args;
+            CORINFO_ARG_LIST_HANDLE arg2                = info.compCompHnd->getArgNext(arg1);
+            var_types               argType             = TYP_UNKNOWN;
+            CORINFO_CLASS_HANDLE    argClass            = NO_CLASS_HANDLE;
+            const bool              isSingleStructStore = sig->numArgs == 3;
+
+            if (isSingleStructStore)
+            {
+                CORINFO_ARG_LIST_HANDLE arg3 = info.compCompHnd->getArgNext(arg2);
+                argType = JITtype2varType(strip(info.compCompHnd->getArgType(sig, arg3, &argClass)));
+                op3     = impPopStack().val;
+            }
+            else
+            {
+                assert(sig->numArgs == 2);
+            }
 
             argType             = JITtype2varType(strip(info.compCompHnd->getArgType(sig, arg2, &argClass)));
             op2                 = impPopStack().val;
             unsigned fieldCount = info.compCompHnd->getClassNumInstanceFields(argClass);
-            argType             = JITtype2varType(strip(info.compCompHnd->getArgType(sig, arg1, &argClass)));
-            op1                 = getArgForHWIntrinsic(argType, argClass);
 
-            assert(op2->TypeGet() == TYP_STRUCT);
+            if (fieldCount > 1 &&
+                ((intrinsic == NI_AdvSimd_StoreSelectedScalar) || (intrinsic == NI_AdvSimd_Arm64_StoreSelectedScalar)))
+            {
+                // StoreSelectedScalar that stores 2/3/4 vectors
+                CORINFO_CLASS_HANDLE structType;
+                unsigned int         sizeBytes   = 0;
+                CORINFO_FIELD_HANDLE fieldHandle = info.compCompHnd->getFieldInClass(argClass, 0);
+                CorInfoType          fieldType   = info.compCompHnd->getFieldType(fieldHandle, &structType);
+                simdBaseJitType                  = getBaseJitTypeAndSizeOfSIMDType(structType, &sizeBytes);
+                if (simdBaseJitType == CORINFO_TYPE_UNDEF) // the argument is not a vector
+                {
+                    CORINFO_CLASS_HANDLE tmpClass;
+                    simdBaseJitType = strip(info.compCompHnd->getArgType(sig, arg2, &tmpClass));
+
+                    if (simdBaseJitType == CORINFO_TYPE_PTR)
+                    {
+                        simdBaseJitType = info.compCompHnd->getChildType(argClass, &tmpClass);
+                    }
+                }
+            }
+
+            if (op2->TypeGet() == TYP_STRUCT)
```

Comment:
i think it will be good to separate out `*StoreSelectedScalar` and `*StoreVector64` because `StoreVector` doesn't have an equivalent of `fieldCout == 1`, so don't want to penalize it with the extra checks.

Reply:
Cool, I'll separate them.

Reply:
Done